packages/components/src/input-control/reducer/state.ts

```diff
@@ -19,7 +19,7 @@ export interface InputState {
 	value?: string;
 }
 
-export type StateReducer = Reducer< InputState, InputAction >;
+export type StateReducer = Reducer< InputState, InputAction | ControlAction >;
```

Comment:
I can see why you'd ask. Looking at this more closely, given we are okay with the way I've wired up `CONTROL` to be exclusively handled by the base reducer, then I'll say no, and yet it's still not quite right as is. The extending reducers will never receive `CONTROL` actions so `StateReducer` would be too loose a type for those.

<details><summary>I'm thinking we should we keep <code>ControlAction</code> out of both <code>InputAction</code> and <code>StateReducer</code> then make only the base reducer typed to accept it.</summary>

```diff
diff --git a/packages/components/src/input-control/reducer/reducer.ts b/packages/components/src/input-control/reducer/reducer.ts
index 3f3302608d..a5f80026a6 100644
--- a/packages/components/src/input-control/reducer/reducer.ts
+++ b/packages/components/src/input-control/reducer/reducer.ts
@@ -1,7 +1,7 @@
 /**
  * External dependencies
  */
-import type { SyntheticEvent, ChangeEvent, PointerEvent } from 'react';
+import type { SyntheticEvent, ChangeEvent, PointerEvent, Reducer } from 'react';
 
 /**
  * WordPress dependencies
@@ -50,7 +50,7 @@ function mergeInitialState(
  */
 function inputControlStateReducer(
 	composedStateReducers: StateReducer
-): StateReducer {
+): Reducer< InputState, actions.InputAction | actions.ControlAction > {
 	return ( state, action ) => {
 		const nextState = { ...state };
 
@@ -151,7 +151,7 @@ export function useInputControlStateReducer(
 	initialState: Partial< InputState > = initialInputControlState,
 	onChangeHandler: InputChangeCallback
 ) {
-	const [ state, dispatch ] = useReducer< StateReducer >(
+	const [ state, dispatch ] = useReducer(
 		inputControlStateReducer( stateReducer ),
 		mergeInitialState( initialState )
 	);
diff --git a/packages/components/src/input-control/reducer/state.ts b/packages/components/src/input-control/reducer/state.ts
index 53a49aa765..2ca9edaa27 100644
--- a/packages/components/src/input-control/reducer/state.ts
+++ b/packages/components/src/input-control/reducer/state.ts
@@ -6,7 +6,7 @@ import type { Reducer, SyntheticEvent } from 'react';
 /**
  * Internal dependencies
  */
-import type { ControlAction, InputAction } from './actions';
+import type { InputAction } from './actions';
 
 export interface InputState {
 	_event?: SyntheticEvent;
@@ -19,7 +19,7 @@ export interface InputState {
 	value?: string;
 }
 
-export type StateReducer = Reducer< InputState, InputAction | ControlAction >;
+export type StateReducer = Reducer< InputState, InputAction >;
 
 export const initialStateReducer: StateReducer = ( state: InputState ) => state;
 
```
</details>

Reply:
The approach that you propose sounds reasonable — maybe we could experiment with generics in the `StateReducer` type, so that we keep all type defs in the same file?

<details>

<summary>Something like this</summary>

```diff
diff --git a/packages/components/src/input-control/reducer/reducer.ts b/packages/components/src/input-control/reducer/reducer.ts
index 3f3302608d..c41ae85775 100644
--- a/packages/components/src/input-control/reducer/reducer.ts
+++ b/packages/components/src/input-control/reducer/reducer.ts
@@ -50,7 +50,7 @@ function mergeInitialState(
  */
 function inputControlStateReducer(
 	composedStateReducers: StateReducer
-): StateReducer {
+): StateReducer< actions.ControlAction > {
 	return ( state, action ) => {
 		const nextState = { ...state };
 
@@ -151,7 +151,7 @@ export function useInputControlStateReducer(
 	initialState: Partial< InputState > = initialInputControlState,
 	onChangeHandler: InputChangeCallback
 ) {
-	const [ state, dispatch ] = useReducer< StateReducer >(
+	const [ state, dispatch ] = useReducer(
 		inputControlStateReducer( stateReducer ),
 		mergeInitialState( initialState )
 	);
diff --git a/packages/components/src/input-control/reducer/state.ts b/packages/components/src/input-control/reducer/state.ts
index 53a49aa765..839ac51383 100644
--- a/packages/components/src/input-control/reducer/state.ts
+++ b/packages/components/src/input-control/reducer/state.ts
@@ -6,7 +6,7 @@ import type { Reducer, SyntheticEvent } from 'react';
 /**
  * Internal dependencies
  */
-import type { ControlAction, InputAction } from './actions';
+import type { InputAction } from './actions';
 
 export interface InputState {
 	_event?: SyntheticEvent;
@@ -19,7 +19,10 @@ export interface InputState {
 	value?: string;
 }
 
-export type StateReducer = Reducer< InputState, InputAction | ControlAction >;
+export type StateReducer< TAdditionalAction = {} > = Reducer<
+	InputState,
+	InputAction | TAdditionalAction
+>;
 
 export const initialStateReducer: StateReducer = ( state: InputState ) => state;
 
```
</details>

We should probably also add a comment about `inputControlStateReducer` being the only reducer exposing / handling the `CONTROL` action — which means that it's an action that doesn't get exposed to external reducer middleware, meant to be used only internally. Maybe in the `inputControlStateReducer`'s JSDocs ?

Reply:
Done in 9aef4eb though it required a touch of the `Action` type as well. At first I didn't favor the idea generalizing the `StateReducer` type for a singular exception to its shape but given that we're interested in refactoring to a hook+component combo https://github.com/WordPress/gutenberg/pull/41264#issuecomment-1139077885 there will be opportunity make more use of the generic.